File: RB101_RB109_small_problems/advanced_1/merge_sort.rb
=begin

PEDAC Template
==============

(Understand the) Problem
------------------------

-  Inputs: array containing elements of a single data type (either numbers or strings)
-  Output: a sorted version of the input array

---

**Problem Domain:**

Merge sort is a recursive sorting algorithm that works by breaking down the array elements into nested sub-arrays, then recombining those nested sub-arrays into sorted order.
Ex:
Breaking down:
[9, 5, 7, 1] ->
[[9, 5], [7, 1]] ->
[[[9], [5]], [[7], [1]]]
Merging each pair of nested sub-arrays and sorting.
[[[9], [5]], [[7], [1]]] ->
[[5, 9], [1, 7]] ->
[1, 5, 7, 9]

---

**Explicit Requirements:**
- Assume the input array has elements all of a single data type
- Assume the input array contains only numbers or strings
- Can use merge sort to sort the input array.
---


**Implicit Requirements:**
- Method must be named #merge_sort
---

**Clarifying Questions:**

1. Should the input array be mutated by the #merge_sort method?
  No. Do not mutate the input array.

---

**Mental Model:**

Breaking down:
  Partition the first and second half of the input array into separate subarrays. Continue until all subarrays are of size 1. Do not partition arrays that are already of size 1.


After breaking down step, all elements are in subarrays of size 1 (meaning they are inherently sorted).

Merging & Sorting:
  Combine subarrays (starting with subarrays of size 1 within the same array) using the #merge method to sort them.


---

Examples / Test Cases / Edge Cases
----------------------------------

**Examples:**

merge_sort([9, 5, 7, 1]) == [1, 5, 7, 9]
merge_sort([5, 3]) == [3, 5]
merge_sort([6, 2, 7, 1, 4]) == [1, 2, 4, 6, 7]
merge_sort(%w(Sue Pete Alice Tyler Rachel Kim Bonnie)) == %w(Alice Bonnie Kim Pete Rachel Sue Tyler)
merge_sort([7, 3, 9, 15, 23, 1, 6, 51, 22, 37, 54, 43, 5, 25, 35, 18, 46]) == [1, 3, 5, 6, 7, 9, 15, 18, 22, 23, 25, 35, 37, 43, 46, 51, 54]

---

Data Structure
--------------
Ex: [9, 5, 7, 1] 

Breaking down:
[9, 5, 7, 1] ->
[[9, 5], [7, 1]] ->
[[[9], [5]], [[7], [1]]]
Merging each pair of nested sub-arrays and sorting:
[[[9], [5]], [[7], [1]]] ->
[[5, 9], [1, 7]] ->
[1, 5, 7, 9]


Ex: [6, 2, 7, 1, 4]

Breaking down:
[6, 2, 7, 1, 4] ->
[[6, 2], [7, 1, 4]] ->
[[[6], [2]], [[7], [1, 4]]] ->
[[[6], [2]], [[7], [[1], [4]]]]
Merging each pair of nested sub-arrays and sorting:
[[[6], [2]], [[7], [[1], [4]]]] ->
[[[6], [2]], [[7], [1, 4]]] ->
[[2, 6], [1, 4, 7]] ->
[1, 2, 4, 6, 7]

---

Algorithm
---------

Breaking down:
- Determine the size of the input array. Assign to variable `size`.
- If `size` equals 1, return the input array. ##This is the break condition out of the recursive loop!
- Split the input array into the first half (from index 0 to `size` / 2 - 1) and second half (from index `size` / 2 to `size` - 1 ).
- Repeat above steps until the input arrays are only of size 1.

Merging and sorting AFTER all subarrays are size 1:
- Merge sorted arrays returned from the recursive break down above. Use the #merge method from the previous exercise.

Code
----

=end

def merge(array1, array2)
  index2 = 0
  result = []

  array1.each do |value|
    while index2 < array2.size && array2[index2] <= value
      result << array2[index2]
      index2 += 1
    end
    result << value
  end

  result.concat(array2[index2..-1])
end

def merge_sort(arr)
  
  #Breaking down steps
  size = arr.size
  #If the input arrays is of size 1, then it is already sorted.
  return arr if size == 1
  
  first_half_arr = arr[0...(size/2)]
  second_half_arr = arr[(size/2)...size]
  
  first_half_arr = merge_sort(first_half_arr)
  second_half_arr = merge_sort(second_half_arr)
  
  #At this point, first_half_arr and second_half_arr are sorted arrays.
  
  #Merging steps using returned, sorted arrays first_half_arr and second_half_arr
  merge(first_half_arr, second_half_arr)
end


p merge_sort([9, 5, 7, 1]) == [1, 5, 7, 9]
p merge_sort([5, 3]) == [3, 5]
p merge_sort([6, 2, 7, 1, 4]) == [1, 2, 4, 6, 7]
p merge_sort(%w(Sue Pete Alice Tyler Rachel Kim Bonnie)) == %w(Alice Bonnie Kim Pete Rachel Sue Tyler)
p merge_sort([7, 3, 9, 15, 23, 1, 6, 51, 22, 37, 54, 43, 5, 25, 35, 18, 46]) == [1, 3, 5, 6, 7, 9, 15, 18, 22, 23, 25, 35, 37, 43, 46, 51, 54]






def merge_sort(array)
  merge_pairs(divide(array))
end

def divide(array)
  if array.size <= 1
    array
  else
    middle_index = array.length / 2
    [divide(array[0...middle_index]), divide(array[middle_index..-1])]
  end
end

p divide([9, 5, 7, 1]) == [[[9], [5]], [[7], [1]]] # Test case

def merge_pairs(array)
  if array.size <= 1
    array
  else
    merge(merge_pairs(array.first), merge_pairs(array.last))
  end
end

p merge_pairs([[[9], [5]], [[7], [1]]]) == [1, 5, 7, 9] # Test case

p merge_sort([5, 3, 4, 8, 1])

File: JS210_small_problems/advanced_1/merge_sort.js
"use strict";
/*

Merge sort is a recursive sorting algorithm that works by breaking down an array's elements into nested subarrays, then combining those nested subarrays back together in sorted order. 

It is best explained with an example. Given the array [9, 5, 7, 1], let's walk through the process of sorting it with merge sort. We'll start off by breaking the array down into nested subarrays:

[9, 5, 7, 1] -->
[[9, 5], [7, 1]] -->
[[[9], [5]], [[7], [1]]]

We then work our way back to a flat array by merging each pair of nested subarrays back together in the proper order:

[[[9], [5]], [[7], [1]]] -->
[[5, 9], [1, 7]] -->
[1, 5, 7, 9]

Write a function that takes an array argument and returns a new array that contains the values from the input array in sorted order. 
The function should sort the array using the merge sort algorithm as described above. You may assume that every element of the array will be of the same data type—either all numbers or all strings.

Feel free to use the merge function you wrote in the previous exercise.


PEDAC Template
==============

(Understand the) Problem
------------------------

-  Inputs: a flat array containing elements of all the same data type (either all numbers or all strings)
-  Output: a new flat array with the elements from the input array sorted using the merge sort algorithm

---

**Problem Domain:**

Merge sort is a *recursive* algorithm. It breaks an array's elements into nested subarrays until each subarray contains only a single element. It then combined the nested subarrays together in sorted order.

Remember, recursive functions must have:
  - A function call to itself
  - A break condition (AKA an endpoint that causes all recursive calls to "unravel")
  - A return value that somehow includes the result of the recursive call.

---

**Explicit Requirements:**
- We must return a new array.
- We can use the #merge function we wrote in the previous exercise.
- The mergeSort function must be a recursive function.
---


**Implicit Requirements:**
- The function must be named #mergeSort
---

**Clarifying Questions:**

1. None at this time.
2.
3.

---

**Mental Model:**

#mergeSort
- If the input array only has a single element, return the input array.
- Else split the input array into two halves.
- Recursively call mergeSort on each half.
- Return the result of using the #merge function on the returned arrays from the mergeSort calls.

---

Examples / Test Cases / Edge Cases
----------------------------------

**Examples:**

mergeSort([9, 5, 7, 1]);           // [1, 5, 7, 9]
mergeSort([5, 3]);                 // [3, 5]
mergeSort([6, 2, 7, 1, 4]);        // [1, 2, 4, 6, 7]
// [6, 2, 7, 1, 4] ->
// [[6, 2, 7] , [1, 4]] ->
// [[[6, 2], [7]] , [[1], [4]]] ->
// [[[[6], [2]], [7]] , [[1], [4]]]  ----->

// [[[2, 6], [7]] , [1, 4]] ->
// [[2, 6, 7] , [1, 4]] ->
// [1, 2, 4, 6, 7] ->

mergeSort(['Sue', 'Pete', 'Alice', 'Tyler', 'Rachel', 'Kim', 'Bonnie']);
// ["Alice", "Bonnie", "Kim", "Pete", "Rachel", "Sue", "Tyler"]

mergeSort([7, 3, 9, 15, 23, 1, 6, 51, 22, 37, 54, 43, 5, 25, 35, 18, 46]);
// [1, 3, 5, 6, 7, 9, 15, 18, 22, 23, 25, 35, 37, 43, 46, 51, 54]

---

Data Structure
--------------

Use arrays to manage the individual elements.

---

Algorithm
---------

#mergeSort
- If the input array only has a single element, return the input array.
- Else split the input array into two halves.
- Recursively call mergeSort on each half.
- Return the result of using the #merge function on the returned arrays from the mergeSort calls.

Code
----

*/

function mergeSort(inputArr) {
  if (inputArr.length === 1) return inputArr;

  let firstHalf = inputArr.slice(0, inputArr.length / 2);
  let secondHalf = inputArr.slice(inputArr.length / 2);

  return merge(mergeSort(firstHalf), mergeSort(secondHalf));
}


function merge(arr1, arr2) {
  if (arr1.length === 0) {
    return arr2.slice();
  } else if (arr2.length === 0) {
    return arr1.slice();
  }

  let mergedAndSorted = [];
  let arr1Copy = arr1.slice();
  let arr2Copy = arr2.slice();

  while (arr1Copy.length > 0 || arr2Copy.length > 0) {
    if (arr1Copy.length === 0) {
      mergedAndSorted.push(arr2Copy.shift());
      continue;
    } else if (arr2Copy.length === 0) {
      mergedAndSorted.push(arr1Copy.shift());
      continue;
    }

    if (arr1Copy[0] <= arr2Copy[0]) {
      mergedAndSorted.push(arr1Copy.shift());
    } else {
      mergedAndSorted.push(arr2Copy.shift());
    }
  }

  return mergedAndSorted;
}

console.log(mergeSort([9, 5, 7, 1]));           // [1, 5, 7, 9]
console.log(mergeSort([5, 3]));                 // [3, 5]
console.log(mergeSort([6, 2, 7, 1, 4]));        // [1, 2, 4, 6, 7]

console.log(mergeSort(['Sue', 'Pete', 'Alice', 'Tyler', 'Rachel', 'Kim', 'Bonnie']));
// ["Alice", "Bonnie", "Kim", "Pete", "Rachel", "Sue", "Tyler"]

console.log(mergeSort([7, 3, 9, 15, 23, 1, 6, 51, 22, 37, 54, 43, 5, 25, 35, 18, 46]));
// [1, 3, 5, 6, 7, 9, 15, 18, 22, 23, 25, 35, 37, 43, 46, 51, 54]